_posts/2021-12-29-readable-tests-by-example.md

```diff
@@ -0,0 +1,550 @@
+            assert event.delivery().cost() == Money.ZERO
+            assert event.when() == CURRENT_DATE
+        })
+}
+```
+
+In this case, it may seem that test is much better, because it is simpler and easier to read, but with a more in-depth
+analysis it turns out that it still does not meet the expected requirements as in the section “How can we test such
+code?”, because it:
+
+* uses concepts such as repository, provider, event publisher, handler, which were not mentioned in the written business
+  scenarios - these are technical implementation details;
+* all the variables used are within the specification, which in the case of a multitude of tests may constitute
+  additional complexity in their maintenance. Besides, it is not difficult to make a mistake here, e.g. by introducing a
+  new variable that is already defined somewhere under a different name;
+* is not easy to add further tests outside the specification that would need similar functionality - unfortunately, it
+  will require multiple repetition of fragments of the code.
+
+I have used two above examples of tests (acceptance and unit), to quickly highlight how many flaws the naive solution
+has adopted, even though the business requirement was not too complicated. In summary, in each of the tests we have had
+to take extra care of:
+
+* manually creating objects using a constructor is not comfortable and additionally, with the large number of
+  parameters, difficult to read; this also makes the tests messy and hard to maintain because changing the constructor
+  makes them very fragile;
+* creating body http requests using text blocks, which in the case of larger objects leads to the creation of structures
+  occupying a large part of the specification;
+* mocking or stubbing external dependencies using mechanisms from the framework as Stub or Mock, which can be
+  comfortable but not necessarily improve the readability of the code and its further development;
+* stubbing the response to external services using the library Wiremock class directly in your code test;
+* checking the final state of an object by referring directly to its content in the test. With complex structures it can
+  be very inconvenient and unreadable.
+
+In the next section, I will focus on eliminating these shortcomings with a few simple solutions.
+
+## Fixing the state of affairs
+
+Let's look at the first test again, which was presented in the previous section “Naive Approach”. It is not too hard to
+notice that the vocabulary here resembles a more natural language, used by domain experts who do not use purely
+technical terms.
+
+```groovy
+class OrderPaymentAcceptanceSpec extends BaseIntegrationTest implements
+    CreateOrderAbility,
+    ClientReputationAbility,
+    SpecialPriceProviderAbility,
+    CourierSystemAbility,
+    OrderPaymentAbility,
+    FreeMusicTrackSenderAbility {
+
+    def "shouldn't charge for delivery when the client has a VIP status"() {
```

Comment:
#minor
Maybe status is not the best word here. The "reputation" word seems to be more elegant. 
People could associate this with some database table column.

Reply:
Since I see no further comments and this has an "outdated" label, marking as resolved.